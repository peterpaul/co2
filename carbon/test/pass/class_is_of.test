/*
 * carbon Compiler and programming language for libco2
 *
 * Copyright (C) 2011,2012 Peterpaul Taekele Klein Haneveld
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
["stdio.h"] {
	int printf(char *format, ...);
}

["co2/Object.h"] {
	class Object {
		void *class;
	}
}

class A : Object {
}

class B : A {
}

int main() {
	A a = new A();
	B b = new B();

	printf ("a is_of B, expected: 0, actual: %d\n", a is_of B);
	printf ("b is_of A: expected: 1, actual: %d\n", b is_of A);

	printf ("a is_of b.class: expected: 0, actual: %d\n", a is_of b.class);
	printf ("b is_of a.class: expected: 1, actual: %d\n", b is_of a.class);

	printf ("!a is_of B, expected: 1, actual: %d\n", !a is_of B);

	return 0;
}
