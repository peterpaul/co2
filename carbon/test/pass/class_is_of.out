a is_of B, expected: 0, actual: 0
b is_of A: expected: 1, actual: 1
a is_of b.class: expected: 0, actual: 0
b is_of a.class: expected: 1, actual: 1
!a is_of B, expected: 1, actual: 1
